lib/pbench/server/database/models/datasets.py

```diff
@@ -730,8 +885,165 @@ def create(**kwargs) -> "Metadata":
+
+        Returns:
+            True if the path is valid, or False
+        """
+        k = key.lower()
+        # Check for exact match
+        if k in valid:
+            return True
+        path = k.split(".")
+        # Disallow ".." and trailing "."
+        if "" in path:
+            return False
+        # Check for open namespace match
+        if path[0] + ".*" not in valid:
+            return False
+        # Check that all open namespace keys are valid symbols
+        return bool(re.fullmatch(Metadata._valid_key_charset, k))
+
+    @staticmethod
+    def getvalue(dataset: Dataset, key: str) -> JSON:
+        """
+        Returns the value of the specified key, which may be a dotted
+        hierarchical path (e.g., "server.deleted").
+
+        The specific value of the dotted key path is returned, not the top
+        level Metadata object. The full JSON value of a top level key can be
+        acquired directly using `Metadata.get(dataset, key)`
+
+        E.g., for "user.contact.name" with the dataset's Metadata value for the
+        "user" key as {"contact": {"name": "Dave", "email": "d@example.com"}},
+        this would return "Dave", whereas Metadata.get(dataset, "user") would
+        return the entire user key JSON, such as
+        {"user" {"contact": {"name": "Dave", "email": "d@example.com}}}
+
+        Args:
+            dataset: associated dataset
+            key: hierarchical key path to fetch
+
+        Returns:
+            Value of the key path
+        """
+        if not Metadata.is_key_path(key, Metadata.METADATA_KEYS):
+            raise MetadataBadKey(key)
+        keys = key.lower().split(".")
+        native_key = keys.pop(0)
+        try:
+            meta = Metadata.get(dataset, native_key)
+        except MetadataNotFound:
+            return None
+        value = meta.value
+        name = native_key
+        for i in keys:
+            # If we have a nested key, and the `value` at this level isn't
+            # a dictionary, then the `getvalue` path is inconsistent with
+            # the stored data; let the caller know with an exception.
+            if type(value) is not dict:
+                raise MetadataBadStructure(dataset, key, name)
+            if i not in value:
+                return None
+            value = value[i]
+            name = i
+        return value
+
+    @staticmethod
+    def setvalue(dataset: Dataset, key: str, value: Any) -> "Metadata":
+        """
+        Create or modify an existing metadata value. This method supports
+        hierarchical dotted paths like "dashboard.seen" and should be used in
+        most contexts where client-visible metadata paths are used.
+
+        This will create a nested JSON structure (represented as a Python dict)
+        as necessary as it descends the hierarchy. For example, if you assign
+        "a.b.c" with an initial value of "{}", you'll end up with
+        {"a": {"b": {"c": value}}}
```

Comment:
{"a": {"b": {"c": {}}}}  ?

Reply:
Slightly rewritten, as I decided it wasn't the best example or explanation ... I had to rebase anyway, so why not?